jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentTempDirTest.java

```diff
@@ -0,0 +1,231 @@
+import org.eclipse.jetty.toolchain.test.FS;
+import org.eclipse.jetty.toolchain.test.MavenTestingUtils;
+import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.Scanner;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.webapp.WebAppContext;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertNotSame;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class DeploymentTempDirTest
+{
+    private static final Logger LOG = Log.getLogger(DeploymentTempDirTest.class);
+
+    private final WebAppProvider webAppProvider = new WebAppProvider();
+    private final ContextHandlerCollection contexts = new ContextHandlerCollection();
+    private final Path testDir = MavenTestingUtils.getTargetTestingPath(DeploymentTempDirTest.class.getSimpleName());
+    private final Path tmpDir = testDir.resolve("tmpDir");
+    private final Path webapps = testDir.resolve("webapps");
+    private final Server server = new Server();
+
+    @BeforeEach
+    public void setup() throws Exception
+    {
+        ServerConnector connector = new ServerConnector(server);
+        server.addConnector(connector);
+
+        FS.ensureEmpty(testDir);
+        FS.ensureEmpty(tmpDir);
+        FS.ensureEmpty(webapps);
+
+        webAppProvider.setMonitoredDirName(webapps.toString());
+        DeploymentManager deploymentManager = new DeploymentManager();
+        deploymentManager.addAppProvider(webAppProvider);
+        server.addBean(deploymentManager);
+
+        HandlerCollection handlerCollection = new HandlerCollection();
+        handlerCollection.addHandler(contexts);
+        handlerCollection.addHandler(new DefaultHandler());
+        deploymentManager.setContexts(contexts);
+        server.setHandler(handlerCollection);
+    }
+
+    @AfterEach
+    public void stop() throws Exception
+    {
+        server.stop();
+    }
+
+    @Test
+    public void testTmpDirectory() throws Exception
+    {
+        Path warPath = MavenTestingUtils.getTestResourcePath("webapps/foo-webapp-1.war");
+        String deploymentXml = "<Configure class=\"org.eclipse.jetty.webapp.WebAppContext\">\n" +
+            "<Set name=\"war\">" + warPath + "</Set>\n" +
+            "<Set name=\"tempDirectory\">" + tmpDir + "</Set>\n" +
+            "<Set name=\"persistTempDirectory\">false</Set>\n" +
+            "</Configure>";
+
+        server.start();
+        WaitScannerListener listener = new WaitScannerListener();
+        webAppProvider.addScannerListener(listener);
+
+        // Add the webapp xml which will will be detected after scan.
+        listener.reset();
+        createNewFile(webapps, "foo-webapp.xml", deploymentXml);
+        webAppProvider.scan();
```

Comment:
Why are there explicit calls to scan()? The DeploymentManager was added to the server, and the WebAppProvider added to it, so both will start when the server starts and start scanning at the default interval of 10. You should either set the interval to 0 and do your own calls to scan() or remove them and change the scan interval to what you want.

Also, I don't  think you should rely on the order of the scanner listeners: it could be the case that your listener gets called back before the WebAppProvider's listener - I don't think the scanner makes any promises as to the ordering of the calls (although from code inspection you can see what is implemented, but I think that's a fragile way to do it).

Reply:
I will change the interval to 0 as I don't want to wait for the automatic scan. The test isn't relying on the order of the listeners, the test calls scan manually so assuming the listeners aren't notified in a separate thread then it is all good.

I just added the test with the listener as a debugging thing to ensure the listeners were actually called. Now that I fixed the problem I can remove it and the test will remain the same without it.